### PR TITLE
Add article "Architecture verification and documentation with jQAssistant and arc42"

### PR DESCRIPTION
**Pull request by stmu-zuhlke**

### CODE REVIEW

File: drafts/architecture-verification-with-jqassistant.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+The sky's the limit! 🚀
+
+## Living Documentation
+
+As mentioned earlier, jQAssistant rules can be utilized for reporting and can enhance existing documentation with up-to-date information. This makes it ideal for creating living documentation that reflects the current state of the system and the relationship between design and implementation.
+
+For reporting, we'll use the [jqassistant-asciidoctor-extension](https://github.com/jqassistant-tooling/jqassistant-asciidoctorj-extensions), which is an extension for asciiDoctor parsing the jQAssistant report and providing inculde directives for the results.
+As a documentation template, we’ll employ [arc42](https://arc42.org), a framework for documenting software architectures. It already provides a structure we can hook in and is available as [asciiDoc template](https://arc42.org/download#file-based-formats). 
+
+You'll find the documentation in the `src/docs` directory of our sample project. TThe project is pre-configured with the Asciidoc Maven plugin and the Asciidoc multipage HTML plugin, which creates a beautiful multipage documentation in your  `target/docs` folder, when running the project with `mvn verify`.
```
Comment (stse) [inline drafts/architecture-verification-with-jqassistant.md:187]: Spelling: TThe -> The

File: drafts/architecture-verification-with-jqassistant.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+- Verify, if the ADR is backed with at least one jQAssistant constraint
+
+To document ADRs we want to use asciidoc files. Unfortunately, there is no jQAssistant plugin out there, which scans adoc files and adds them to the tree. But there are already other scanners included in jQAssistant, e.g. xml, json and yaml scanners. So as a workaround, we split up the ADR definition in two parts:
+
+1. **ADR yaml file**, implementing an ADR template and providing a unique ADR ID.
+2. **ADR adoc documentation file**, includes the yaml content and the jQAssitant rules for this ADR.
+
+In our sample project you'll find the ADR examples in the folder `src/docs/adr`.
+
+With the yaml files ingested by jQAssistant we can then write concepts and constraints targeting those nodes. Finally we can include all ADR documentation files in the architecture design desciions document.
```
Comment (stse) [inline drafts/architecture-verification-with-jqassistant.md:245]: Spelling desciions

**Review by stse — COMMENTED** [review]
I like it. At the end you are talking about the steep learning curve. I fully agree with that. On suggestion from my side would therefore be to maybe also explain how to execute the rules and how those rules are basically structured and how they work, butI haven't looked into the example project. Maybe this would answer my questions.

**Comment by stse** [comment]
Another thing I usually struggling with is how to identify the stereotype e.g. is it an entity, an adapter, a port an infrastructure level component, an application service,... of a class if this is not encode in the package structure. For example it is usual in a DDD approach to structure your application based on the bounded contexts and aggregate roots which means you package name looks similar to lab.<application>.<context/aggregate root> and everything, e.g. repository events services,... is included in this package. Packages are cut according to domain aspects not to technical aspects. In this case I need something else to identify the stereotype of a class. Any Idea?

**Comment by stmu-zuhlke** [comment]
> Another thing I usually struggling with is how to identify the stereotype e.g. is it an entity, an adapter, a port an infrastructure level component, an application service,... of a class if this is not encode in the package structure. For example it is usual in a DDD approach to structure your application based on the bounded contexts and aggregate roots which means you package name looks similar to lab..<context/aggregate root> and everything, e.g. repository events services,... is included in this package. Packages are cut according to domain aspects not to technical aspects. In this case I need something else to identify the stereotype of a class. Any Idea?

You are right, the given sample project is quiet simple and represents only one context. This makes it easy to identify the concepts based on the conventions (but this was intentional to keep it simple here :D )
with a multi context setup you could still use the depth of packages and then go deeper there (cypher provides some string functionality and with ASOP.
There is always some convention to identify it, be it naming conventions, package conventions or you could introduce your own annotations for that. If there is none, you should create one to make concepts explicit and visible.

Basically this is about testability of the architecture. We already do it with code : make it small, make it independent, if you cannot write a test for it there's something fishy. Same goes for architecture: if you cannot test your implementation structure easy enough, maybe something should be more visible or also amybe something is not there at all.

**Comment by stmu-zuhlke** [comment]
> I like it. At the end you are talking about the steep learning curve. I fully agree with that. On suggestion from my side would therefore be to maybe also explain how to execute the rules and how those rules are basically structured and how they work, butI haven't looked into the example project. Maybe this would answer my questions.

Thanks  :)
Actually I started over multiple times, diving into the "how" just to realise that I basically re-sample blog entries and manuals which are already there and doe it way better (e.g. the link in the introduction). That's the reason why I concentrate more on the "why" and bring up some ideas how to actually use this tool

File: drafts/architecture-verification-with-jqassistant.md [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+            RETURN file
+    ]]></cypher>
+    </constraint>
+```
+
+### Reporting and documentation
+
+With everything in place, we are now ready to incorporate the results into the documentation..
+As you may recall, our objective is to provide a comprehensive overview of all ADRs and their associated constraints.
+o facilitate this, we are introducing a new concept called `adr:constraintReport`. This concept does not modify the graph in any way. Rather we use the query result as a structured report.
```
Comment (immo-huneke-zuhlke) [inline drafts/architecture-verification-with-jqassistant.md:372]: Missing "T" at start of line

**Comment by stmu-zuhlke** [comment]
> Great article. I noticed a couple of spelling errors, which I have noted in the text.
> 
> There's a certain barrier to adoption, which is the very technical syntax of the XML rules file(s). I wonder whether this might be usefully overcome by using Anthony Anjorin's Pedantic extension to MPS - this is a model-building tool par excellence, which may be able to generate the specification language without requiring users to learn its syntax.
> 
> Anthony is a member of our Global CDQ practice. His github repo can be found under https://github.com/anthonyanjorin

This sound like an interesting approach. I am not really fluent with DSLs and MPS but will have a look into this.
However, from author's viewpoint I find it interesting that you see the xml as a possible barrier, since the main implementation of the rules is written in Cypher (the query language from neo4j). The xml is just metadata and orchestration.
I think I will remove the xml part from most of the code blocks to make this more clear and also add a section about the workflow how such rules can be developed using the neo4j tools.
Because what you are doing in the end is, creating your own "domain" graph within neo4j and query it.

**Comment by immo-huneke-zuhlke** [comment]
> This sound like an interesting approach. I am not really fluent with DSLs and MPS but will have a look into this. However, from author's viewpoint I find it interesting that you see the xml as a possible barrier, since the main implementation of the rules is written in Cypher (the query language from neo4j). The xml is just metadata and orchestration.

Thanks for clarifying. I really meant the parts embedded in the XML as CDATA. I would imagine that there's a lot of scope for subtle errors if this is edited directly. Of course, if you're using another tool to develop the Cypher statements this is a different matter (particularly if you can unit-test the query language statements somehow).

**Comment by stmu-zuhlke** [comment]
I did some restructuring and added some more infos about the "how to debug"

**Review by immo-huneke-zuhlke — APPROVED** [review]
It really has become much clearer with these changes and updates. A general spell check would be a good idea before release.

File: drafts/architecture-verification-with-jqassistant.md [hunk 4]
```diff
@@ -322,24 +312,20 @@ jqassistant:
...
-        <cypher><![CDATA[
-            match (adr:Adr)
-            match (e:Xml:Element {name: 'constraint'})-->(attr:Xml:Attribute {name: 'id'})
-            where attr.value STARTS WITH adr.adrId
-            merge (adr)-[r:ENSURED_BY]->(constraint:AdrConstraint {adrId: adr.adrId, id: attr.value})
-            return adr.adrId, type(r), constraint.id
-        ]]></cypher>
-    </concept>
+```Cypher
+// Concept: adr:matchingConstraint
```
[the remark this answers is not in the document]
Comment (stmu-zuhlke) [reply drafts/architecture-verification-with-jqassistant.md:316]: actually that'S a good finding of inconsistency. I will change the label

File: drafts/architecture-verification-with-jqassistant.md [hunk 5]
```diff
@@ -235,14 +235,11 @@ Here is what we want to do:
...
-1. **ADR yaml file**, implementing an ADR template and providing a unique ADR ID.
-2. **ADR adoc documentation file**, includes the yaml content and the jQAssitant rules for this ADR.
+To document ADRs we use asciidoc files. For scanning, we will use this [Asciidoc Scanner Plugin](https://github.com/kontext-e/jqassistant-asciidoc-plugin). The ADR files follow a template, starting with a table holding the header information like a unique ID, the author, and the status.
 
 In our sample project you'll find the ADR examples in the folder `src/docs/adr`.
 
-With the yaml files ingested by jQAssistant we can then write concepts and constraints targeting those nodes. Finally we can include all ADR documentation files in the architecture design desciions document.
+With the asciidoc files ingested by jQAssistant we can then write concepts and constraints targeting those nodes. Finally we can include all ADR documentation files in the architecture design decisions document.
 
 So this is the concept:
```
[the remark this answers is not in the document]
Comment (stmu-zuhlke) [reply drafts/architecture-verification-with-jqassistant.md:245]: I fixed all of those now and ran a spell checker with no further results :) thanks a lot
